// sdutils/src/index.js
/**
 * @file 入口，每个工具方法在单独的文件中，通过入口组合输出
 * @author sensorsdata.cn
 */

// 以浅拷贝的方式展开对象
// 将数组转为以指定字段为 key 的 Object
import arrayToObject from './arrayToObject';
// 检查规则
import checkRules from './checkRules';
// cron 表达式相关方法
import cron from './cron';
// 比较两个值是否全等
import diff from './diff';
// 格式化属性下拉数据
import formatProperties from './formatProperties';
// 为数字添加千位分隔符
import numberSeparator from './numberSeparator';
// 格式化权限方法工厂
import permissionCreator from './permissionCreator';
// 生成随机字符串
import randomStr from './randomStr';
import spreadObject from './spreadObject';
// localStorage 相关工具方法
import storage from './storage';
// 名称切割方法
import formatTagName from './formatTagName';
// 国际化相关共用方法
import intlUtils from './intl';
// 中文基数转英文序数
import { cardinalToOrdinal, cardinalToShortOrdinal, getEnv } from './intl';

const sdutils = {
  spreadObject,
  arrayToObject,
  diff,
  numberSeparator,
  cron,
  storage,
  randomStr,
  checkRules,
  formatProperties,
  permissionCreator,
  formatTagName,
  cardinalToOrdinal,
  cardinalToShortOrdinal,
  getEnv
}

export { spreadObject, arrayToObject, diff, numberSeparator, cron, storage, randomStr, checkRules, formatProperties, permissionCreator, formatTagName, intlUtils, cardinalToOrdinal, cardinalToShortOrdinal, getEnv };

export default sdutils;
